prow/sidecar/censor.go

```diff
@@ -112,6 +116,204 @@ func (o Options) censor() error {
+			logrus.WithError(err).Warn("Could not close input file while detecting content type.")
+		}
+	}()
+
+	header := make([]byte, 512)
+	if _, err := file.Read(header); err != nil {
+		return "", fmt.Errorf("could not read file to check content type: %v", err)
+	}
+	return http.DetectContentType(header), nil
+}
+
+// handleArchive unravels the archive in order to censor data in the files that were added to it.
+// This is mostly stolen from build/internal/untar/untar.go
+func handleArchive(archivePath string, wg *sync.WaitGroup, censor func(file string)) error {
+	outputDir, err := ioutil.TempDir("", "tmp-unpack")
+	if err != nil {
+		return fmt.Errorf("could not create temporary dir for unpacking: %w", err)
+	}
+
+	defer func() {
+		if err := os.RemoveAll(outputDir); err != nil {
+			logrus.WithError(err).Warn("Failed to clean up temporary directory for archive")
+		}
+	}()
+
+	if err := unarchive(archivePath, outputDir); err != nil {
+		return fmt.Errorf("could not unpack archive: %w", err)
+	}
+
+	children := &sync.WaitGroup{}
+	if err := filepath.Walk(outputDir, func(absPath string, info os.FileInfo, err error) error {
+		if info.IsDir() {
+			return nil
+		}
+
+		wg.Add(1)
+		children.Add(1)
+		go func() {
+			censor(absPath)
+			children.Done()
+		}()
+		return nil
+	}); err != nil {
+		return fmt.Errorf("could not walk unpacked archive to censor them: %w", err)
+	}
+
+	children.Wait()
+	if err := archive(outputDir, archivePath); err != nil {
+		return fmt.Errorf("could not re-pack archive: %w", err)
+	}
+	return nil
+}
+
+// unarchive unpacks the archive into the destination
+func unarchive(archivePath, destPath string) error {
+	input, err := os.Open(archivePath)
+	if err != nil {
+		return fmt.Errorf("could not open archive for unpacking: %w", err)
+	}
+	zipReader, err := gzip.NewReader(input)
+	if err != nil {
+		return fmt.Errorf("could not read archive: %w", err)
+	}
+	tarReader := tar.NewReader(zipReader)
+	defer func() {
+		if err := zipReader.Close(); err != nil {
+			logrus.WithError(err).Warn("Could not close zip reader after unarchiving.")
+		}
+		if err := input.Close(); err != nil {
+			logrus.WithError(err).Warn("Could not close input file after unarchiving.")
+		}
+	}()
+
+	for {
+		entry, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("could not read archive: %w", err)
+		}
+		if !validRelPath(entry.Name) {
+			return fmt.Errorf("tar contained invalid name error %q", entry.Name)
+		}
+		rel := filepath.FromSlash(entry.Name)
+		abs := filepath.Join(destPath, rel)
+		mode := entry.FileInfo().Mode()
+		switch {
+		case mode.IsDir():
+			if err := os.MkdirAll(abs, 0755); err != nil {
+				return fmt.Errorf("could not create directory while unpacking archive: %w", err)
+			}
+		case mode.IsRegular():
+			file, err := os.OpenFile(abs, os.O_RDWR|os.O_CREATE|os.O_TRUNC, mode.Perm())
+			if err != nil {
+				return err
+			}
+			n, err := io.Copy(file, tarReader)
```

Comment:
Any idea how the memory consumption of this looks like? Depending on that it might be worth parallellizing, since CPU is a compressible resource so grabbing as much as we can helps. Also I imagine it includes a lot of syscall, which would be another good reason to parallelize

Reply:
IIRC the internal buffers are pretty small for `io.Copy()` between two `os.File` - so it's efficient in space. Yes, we could make the file writes out parallel